Closes #807, Closes #1987, #2446 Switch to CKEditor5 and update text editor and format settings.

joeparsons opened the pull request:
## Description

Alternate CKEditor5 PR that _does not_ move `az_barrio`'s Arizona Bootstrap theme settings to `az_core` like #2596 does.

- Upgrades `az_standard` text format and editor configs to CKEditor5
- Upgrades `basic` text format and editor configs to CKEditor5
- Installs CKEditor5 module instead of CKEditor (4) module by default on new site installs
- Adds DB update to install CKEditor5 module on existing sites
- Updates icons for custom embed buttons to SVGs (needed for CKEditor5)
- Adds CKEditor5 stylesheets to AZ Barrio theme
- Moves most of `az_barrio`'s `hook_library_info_alter()` implementation (and associated helper functions) to `az_core` so AZ Barrio libraries (including Arizona Bootstrap) are loaded correctly in admin theme

Questions we still need to consider:
- Is it going to be safe for people to apply these config changes via Config Distro if they have customized the text formats or editors in any way?
- When (if ever) do we want to automate the uninstallation of the contrib ckeditor module?
- Should we remove the remaining CKeditor 4 specific things from Quickstart now or later? (AZ Barrio ckeditor styles, `az_media_ckeditor_styles_alter()`, etc)

See validation messages that came up when switching editor/text_format configs to CKeditor5 here: https://github.com/az-digital/az_quickstart/issues/1987#issuecomment-1640724617

## Related issues
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

- Closes #807
- Closes #1987
- #2446

## How to test
https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-2769.probo.build/
Verify that editor works with new and existing content at least as well as it did with CKEditor 4.


## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

joeparsons commented on modules/custom/az_paragraphs/az_paragraphs.libraries.yml:
```diff
@@ -31,16 +29,14 @@ az_paragraphs.az_text_media:
   dependencies:
-    - az_barrio/global-styling
-    - az_barrio/arizona-bootstrap
+    - az_core/arizona-bootstrap
```
```suggestion
    - az_barrio/arizona-bootstrap
```

joeparsons commented on themes/custom/az_barrio/css/ckeditor5.css:
```diff
@@ -0,0 +1,40 @@
+  flex-wrap: wrap;
+}
+
+/* /*
```
```suggestion
/*
```

joeparsons commented on themes/custom/az_barrio/css/ckeditor5.css:
```diff
@@ -0,0 +1,40 @@
+ .ck.ck-content .table {
+  width: auto;
+}
+/*
```
```suggestion
/**
```

sanjanans commented:
I have completed testing the CK editor. Found very few minor issues. None of them should be a blocker. 

However, you may want to take a look at my test page https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-2769.probo.build/sanjana-test. The “text” paragraph type (Table: Testing Embeds) is rendering in a weird way in edit mode. Not sure what caused it. I did try to embed all sorts of stuff in that table. 

Below I’ve listed a few minor issues I discovered:

1. “**Insert code block**” in the CK editor: While adding code, there is no way to click out of it. Of course, there are ways around it, and also there won’t be much need for this feature so this should not be a blocker.
2. In **"Styles": “Dark table header”** does not work, or I do not know how to use it.
3. **Small table** is NOT rendered small on the page though it looks small in edit mode.
4. **Table caption** is vertical in edit mode though it shows properly (horizontal) on the page.
5. Not sure if text alignment and top/bottom margins in “styles” are working. It is possible I do not know the use case for it.

danahertzberg commented:
I recommend having the same order of buttons to help ease the transition.

joeparsons commented on modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml:
```diff
@@ -4,96 +4,277 @@ dependencies:
+      - '|'
+      - alignment
+      - '|'
+      - insertTable
```
```suggestion
      - bootstrapGrid
      - insertTable
```

joeparsons commented on modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml:
```diff
@@ -4,96 +4,277 @@ dependencies:
+      - '|'
+      - code
+      - codeBlock
+      - bootstrapGrid
```
```suggestion
```

joeparsons commented on themes/custom/az_barrio/az_barrio.info.yml:
```diff
@@ -48,6 +48,10 @@ ckeditor_stylesheets:
   - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
 
+ckeditor5-stylesheets:
+  - css/ckeditor-arizona-bootstrap.min.css
+  - css/ckeditor5.css
```
```suggestion
  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
  - css/ckeditor5.css
```

joeparsons commented on az_quickstart.install:
```diff
@@ -118,3 +118,10 @@ function az_quickstart_update_920701() {
+/**
+ * Ensure ckeditor5 module is installed.
+ */
+function az_quickstart_update_920801() {
```
```suggestion
function az_quickstart_update_1020801() {
```

joeparsons commented on themes/custom/az_barrio/az_barrio.info.yml:
```diff
@@ -48,6 +48,11 @@ ckeditor_stylesheets:
   - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
 
+ckeditor5-stylesheets:
+  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
+  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
+  - css/ckeditor5.css
```
```suggestion
  - css/ckeditor5.css
  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
```

joeparsons commented on themes/custom/az_barrio/css/az-barrio-off-canvas-nav.css:
```diff
@@ -6,3 +6,11 @@ header#header_arizona {
 .navbar-offcanvas.no-navigation-region {
   border: 0;
 }
+
+/*
+ * Remove when https://github.com/az-digital/arizona-bootstrap/pull/901 is
+ * included in a release
+ */
+.btn-redbar .icon-text {
+  margin: -5px 0;
+}
```
```suggestion
```

trackleft commented on modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install:
```diff
@@ -5,6 +5,8 @@
  * Install, update and uninstall functions for az_paragraphs_text_media module.
  */
 
+use Drupal\paragraphs\Entity\Paragraph;
```
Was this important?

joeparsons replied:
No.  It was unrelated to the other changes in this PR so I omitted the change.

trackleft commented on themes/custom/az_barrio/az_barrio.info.yml:
```diff
@@ -48,6 +48,11 @@ ckeditor_stylesheets:
+ckeditor5-stylesheets:
+  - css/ckeditor5.css
+  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
+  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
```
```suggestion
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest-2.x/css/arizona-bootstrap.min.css
```

joeparsons replied:
Thanks

trackleft commented on themes/custom/az_barrio/az_barrio.info.yml:
```diff
@@ -48,6 +48,11 @@ ckeditor_stylesheets:
   - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
   - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
```
```suggestion
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest-2.x/css/arizona-bootstrap.min.css
```

trackleft commented on modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml:
```diff
@@ -4,96 +4,277 @@ dependencies:
+      linkit_enabled: true
+      linkit_profile: az_linkit
+    media_media:
+      allow_view_mode_override: true
```
What is this `allow_view_mode_override` setting?

joeparsons replied:
Not sure.  Was automatically added by the editor config upgrade.

joeparsons commented:
Updates since 2023-09-29 Friday Meeting:

- Updated patches for Embed and Entity Embed
- Reordered buttons in AZ Standard editor config to match CKeditor 4
- Replaced experimental build of AZ Bootstrap with latest 2.x version from CDN (basically matching CKEditor 4 functionality for now)
- Removed `hook_library_info_alter()` changes since they don't actually seem needed until we're ready to properly solve #2598 
- Removed a couple of unrelated changes

danahertzberg commented:
Testing as a content editor and the functionality is looking great. Improved UX yet still similar to what was there before.

https://53ebdf30-2d20-43df-bbe3-c19a2c93ce72.probo.build/dana-content-editor

danahertzberg commented:
Not blockers imo. But there are some styles that don't render properly in the styles dropdown/editing field.
- Triangle bullets
- Button outline white (needs background color)
- Text white (needs background color)
- Sanserif text not needed anymore
- Text align top
- Text align middle
- Text align bottom
- Zero margin top
- Zero margin bottom
- Table styles?

sanjanans commented:
There seems to be a slight issue with the bootstrap grid in edit mode. On [Sanjana Test page](https://53ebdf30-2d20-43df-bbe3-c19a2c93ce72.probo.build/sanjana-test) I created a 3 col grid and inserted a video in the 1st column. It took up the space for the 2nd col as well. I had to click inside the 1st grid to add text in the 2nd col.
This issue is only in the admin interface. Once saved, it renders correctly on the page. 
<img width="1046" alt="Screenshot 2023-10-03 at 12 15 08 PM" src="https://github.com/az-digital/az_quickstart/assets/71897492/b6617606-e149-46e1-a2af-69dd35369460">

sanjanans reviewed (approved):
This is ready!! Thumbs up from me.
